FinderCompass/src/main/java/atomicstryker/findercompass/client/ThreadCompassWorker.java

```diff
@@ -1 +1,123 @@
+
+/**
+ * Runnable worker class for finding Blocks
+ * 
+ * @author AtomicStryker
+ */
+
+public class ThreadCompassWorker extends Thread
+{
+	private Minecraft mcinstance;
+	private boolean isRunning = false;
+	
+	public ThreadCompassWorker(Minecraft mc, FinderCompassLogic creator)
+	{
+		mcinstance = mc;
+	}
+	
+	private Set<BlockData> mBlocks;
+	private int[] intArray;
+	private String mOreDictName;
+	
+	public void setupValues(Set<BlockData> pBlocks, int[] configInts, String pOreDictName)
+	{
+	    mBlocks = pBlocks;
+		intArray = configInts;
+		mOreDictName = pOreDictName;
+	}
+	
+	public boolean isWorking()
+	{
+		return isRunning;
+	}
+
+	@Override
+	public void run()
+	{
+		isRunning = true;
+		
+		// search!
+		ChunkCoordinates result = findNearestBlockChunkOfIDInRange(mBlocks, intArray[1], intArray[2], intArray[3], intArray[4], intArray[5], intArray[6], intArray[7]);
+		
+		if (result != null)
+		{
+		    FinderCompassClientTicker.instance.onFoundChunkCoordinates(result, mBlocks, mOreDictName);
+		}
+		
+		isRunning = false;
+	}
+	
+    private ChunkCoordinates findNearestBlockChunkOfIDInRange(Set<BlockData> blocks, int playerX, int playerY, int playerZ, int xzRange, int yRange, int minY, int maxY)
+    {
+        List<ChunkCoordinates> blocksInRange = this.findBlocksOfIDInRange(blocks, playerX, playerY, playerZ, xzRange, yRange, minY, maxY);
+        ChunkCoordinates playerCoords = new ChunkCoordinates(playerX, playerY, playerZ);
+        ChunkCoordinates resultCoords = new ChunkCoordinates(0, 0, 0);
+        double minDist = 9999.0D;
+
+        for (int i = 0; i < blocksInRange.size(); ++i)
+        {
+            ChunkCoordinates coords = (ChunkCoordinates)blocksInRange.get(i);
+            double localDist = this.getDistanceBetweenChunks(playerCoords, coords);
+            if (localDist < minDist)
+            {
+                resultCoords = coords;
+                minDist = localDist;
+            }
+        }
+        
+        //System.out.printf("Compassworker found stuff of id %s at [%d|%d|%d]\n", blockID, resultCoords.posX, resultCoords.posY, resultCoords.posZ);
+        return resultCoords;
+    }
+
+    private List<ChunkCoordinates> findBlocksOfIDInRange(Set<BlockData> blocks, int playerX, int playerY, int playerZ, int xzRange, int yRange, int minY, int maxY)
+    {
+        ArrayList<ChunkCoordinates> resultList = new ArrayList<ChunkCoordinates>();
+
+        for (int yIter = playerY - yRange - 1; yIter <= playerY + yRange; ++yIter)
+        {
+            if (yIter >= minY && yIter <= maxY)
+            {
+                for (int zIter = playerZ - xzRange; zIter <= playerZ + xzRange; ++zIter)
+                {
+                    for (int xIter = playerX - xzRange; xIter <= playerX + xzRange; ++xIter)
+                    {
+                    	for (BlockData blockData : blocks)
+                    	{
+                    		if (this.mcinstance.theWorld.getBlock(xIter, yIter, zIter) == blockData.getBlockID())

```

Comment:
This spot is what i mean by "inner loop". You are re-getting the getBlock result when you should be saving it in a local variable BEFORE iterating "blocks"